iree/hal/vulkan/vulkan_device.cc

```diff
@@ -434,7 +437,8 @@ VulkanDevice::VulkanDevice(
           make_ref<DescriptorPoolCache>(add_ref(logical_device_))),
       dispatch_command_pool_(std::move(dispatch_command_pool)),
       transfer_command_pool_(std::move(transfer_command_pool)),
-      legacy_fence_pool_(std::move(legacy_fence_pool)) {
+      legacy_fence_pool_(std::move(legacy_fence_pool)),
+      renderdoc_capture_manager_(renderdoc_capture_manager) {
```

Comment:
all these should just be a capture manager - not the renderdoc-specific one as that defeats the whole purpose of dependency injection. if you only carried around the capture manager then the app could provide their own.

Reply:
(Just replying, no code changes yet)
Good point, I can change these to the base class.

---

The design question I was grappling with was whether the HAL Driver/Device should control connecting and starting/stopping capture itself, or if that should be left up to the higher level systems (to more closely map to function invocations).

Exposing the base class on hal::Driver (`virtual DebugCaptureManager* debug_capture_manager()`) was a step in the direction of letting higher level systems control it, but I'm thinking of removing it from there or moving that field from `public` to `protected`. I could keep it public if there was a compelling reason, like wanting to call `SetCaptureFileComments()` with information about the program being run, its inputs, etc. Perhaps I could make _that_ public, but make the internals of starting/stopping be internal to the subclasses and `hal::Driver` / `hal::Device`?

The timing required for RenderDoc's setup doesn't map to the high level systems well. Specifically with iree-run-mlir, I was looking at making a change (with API bindings) around here: 
https://github.com/google/iree/blob/25dc6c19ea6e8f83089c90a7fa2911d0a6ac329f/iree/tools/vm_util.cc#L268-L277

when I'd conceptually want to be able to start/stop capture for each function invocation:

https://github.com/google/iree/blob/25dc6c19ea6e8f83089c90a7fa2911d0a6ac329f/iree/tools/run_mlir_main.cc#L247-L250

For the command line / one-shot case, having the `Device` start recording at initialization and stop recording at destruction, then using `vkCmdBeginDebugUtilsLabelEXT` / `vkCmdEndDebugUtilsLabelEXT` (or equivalent) to annotate invocations / semantic groupings of commands seems like it would get us captures in a useful format. I'm not so sure for long-running / in-application uses - maybe the app itself would be able to handle setup?

Reply:
Changed to use the base class in VulkanDevice. I'm still wondering if there's a way I could better structure these classes to bridge between high level function invocations / application APIs and the HAL driver/device internals.

Reply:
Discussed offline a bit. May change in a later PR to expose `CaptureManager` through the API, make the specific timing of connecting to RenderDoc an implementation detail of `RenderDocCaptureManager`, and trigger captures more directly in applications (such as `iree-run-mlir`).